zuul-core/src/main/java/com/netflix/zuul/message/Headers.java

```diff
@@ -195,6 +195,30 @@ public void setAndValidate(String headerName, @Nullable String value) {
         setNormal(validateField(headerName), validateField(normalName), validateField(value));
     }
 
+    /**
+     * Replace any/all entries with this key, with this single entry if the key and entry are valid.
+     *
+     * If value is {@code null}, then not added, but any existing header of same name is removed.
+     */
+    public void setIfValid(HeaderName headerName, String value) {
+        String normalName = requireNonNull(headerName, "headerName").getNormalised();
+        if (isValid(headerName.getName()) && isValid(normalName) && isValid(value)) {
+            setNormal(headerName.getName(), normalName, value);
+        }
+    }
+
+    /**
+     * Replace any/all entries with this key, with this single entry if the key and entry are valid.
+     *
+     * If value is {@code null}, then not added, but any existing header of same name is removed.
+     */
+    public void setIfValid(String headerName, @Nullable String value) {
+        String normalName = HeaderName.normalize(requireNonNull(headerName, "headerName"));
+        if (isValid(headerName) && isValid(normalName) && isValid(value)) {
```

Comment:
nit: the validity checks for names and values are different I thought?   It's important that validity is well defined at the creation of this new API method, because it will become hard to change it later.

Reply:
Ideally header name validation should be stricter, but want to make sure that doing that does not break any existing behavior. Keeping the validation of header names relaxed in this PR for now.